### PR TITLE
webpack needs -p flag in order to minimize files for production

### PR DESCRIPTION
**Pull request by patrickleet**

### CODE REVIEW

**Comment by Bartekus** [comment]
I'm not entirely sure that it doesnt do that already since we do use define-plugin, but certainly it cannot hurt so indeed, good catch @patrickleet , I'll have this implemented asap

**Comment by patrickleet** [comment]
I thought it did as well, until I deployed and ran pagespeed, and it told me my JS and CSS was not minified.

**Comment by Bartekus** [comment]
ahh, that certainly confirms it, good catch @patrickleet I'll have this done right away!

**Comment by Bartekus** [comment]
Tho the problem that I see now is that we might want to create another set of npm scripts to do the final:build that would have the webpack with -p as at the moment when you run dev mode, it will still use production build but if we use -p, that could impede the troubleshooting. Tho no matter what, the final build needs to be produced with -p so you are correct... I need a moment to think about this and look into npm lifecycle... perhaps I'll find an eloquent solution there

**Comment by mattkrick** [comment]
this shouldn't be necessary, a production `NODE_ENV` + config file will give you a production build. Checking out what the client gets, I see all minimized code. Could you offer up steps to show why it isn't minimized?

**Comment by patrickleet** [comment]
Yea, seems necessary. I didn't think so either as at first glance it appeared somewhat minified, but I got more minification out of just including the -p flag.

Try running it both ways and open the js files to compare the difference.

**Comment by Bartekus** [comment]
I'll second this, -p doesn't hurt and certainly forces webpack to be more vigorous with code compacting.

**Comment by mattkrick** [comment]
Definitely agree, I'm just more curious about the why, because if the p arg
does something different than setting the environment it should be a bug in
webpack

**Comment by Bartekus** [comment]
From what I gather and understand, NODE_ENV=production only set process.env.NODE_ENV to production in node environment however being that webpack is also capable of optimizing the frontend code which run on browsers, additional flag has to supersede the entirety of the code scope that webpack will optimize thus -p acts as such measure. However being that babel is also is involved in the process I think one should not skip NODE_ENV=production since that's the flag that babel will use.
So to be more specific here -p forces webpack into these two modes --optimize-minimize & --optimize-occurence-order. In Meatier there is also DefinePlugin, that work as another layer of optimization that allows uglify to simply drop the unused code while preventing webpack from causing things such as: Uncaught ReferenceError: process is not defined(…) This is done by having webpack do something like this to frontend react code:
```
...
if ("production" !== process.env.NODE_ENV) {
...
    ("production" !== process.env.NODE_ENV ? warning(
      standardName == null,
      'Unknown DOM property ' + name + '. Did you mean ' + standardName + '?'
    ) : null);
...
}
...
```
So webpack replaces the top-most if ("production" !== process.env.NODE_ENV) with if (false) and skips the block. Subsequently the process.env.NODE_ENV in the block never gets executed and UglifyJs will remove it as dead code.
```
...
if (false) {
...
    ("production" !== process.env.NODE_ENV ? warning(
      standardName == null,
      'Unknown DOM property ' + name + '. Did you mean ' + standardName + '?'
    ) : null);
...
}
...
```
So in short, the DefinePlugin just replaces occurrences of the given identifiers with the given expressions. After that, UglifyJS detects dead code blocks and removes them.

All this is done because Webpack does all of its AST walking and transforming together on an AST, and it does so while performing static analysis, so if it recognizes that it can't get into one side of an if, it won't run any transforms in that part. DefinePlugin hooks into that transform/traversal so it won't bother substituting stuff into the if(false){... side of the if.

**Comment by mattkrick** [comment]
Got some free time to dig into this!
The culprit is `LoaderOptionsPlugin`:
https://github.com/webpack/webpack/blob/a90b96830261367192dd35d15bf9ffd670b541be/bin/convert-argv.js#L430-L432

the `-p` flag additionally minimizes the loaders, thus the difference.

I think the _right_ way to do this is to add this to the config. For example, if you've got some deploy-only plugins...

```
const deployPlugins = [];
if (process.env.DEPLOY) {
  deployPlugins.push(new webpack.optimize.UglifyJsPlugin({compressor: {warnings: false}, comments: /(?:)/}));
  deployPlugins.push(new webpack.LoaderOptionsPlugin({comments: false}));
}
```
